### PR TITLE
feat: support for specifying port in external services

### PR DESCRIPTION
**Pull request by Gallardot**
Signed-off-by: Gallardot <tttick@163.com>

### Type of change:

- [ ] Bugfix
- [ ] New feature provided
- [ ] Improve performance
- [ ] Backport patches

### Pre-submission checklist:

* [ ] Did you explain what problem does this PR solve? Or what new features have been added?
* [ ] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix-ingress-controller#community) first**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -49,7 +49,7 @@ We have added partial support for Gateway API, which is not enabled by default,
 Ingress resources can now use all APISIX plugin configurations by setting the annotation `k8s.apisix.apache.org/plugin-config-name=xxx`.
 
 Please try out the release binaries and report any issues at
-https://github.com/apache/apisix-ingress-controller/issues.
+<https://github.com/apache/apisix-ingress-controller/issues>.
```
Comment (tao12345666333) [inline CHANGELOG.md:52]: This file is automatically generated by the tool, so there is no need to modify it.

**Review by lingsamuel — APPROVED** [review]
Thanks!

File: CHANGELOG.md (hunk 1 above)
Comment (Gallardot) [reply CHANGELOG.md:52]: This file is modified by the `make update-all` command. Let me see if I can skip it.

Comment (Gallardot) [reply CHANGELOG.md]: @tao12345666333 I've fixed it.

**Review by lingsamuel — APPROVED** [review]

File: pkg/providers/utils/scheme.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+var schemeToPortMaps = map[string]int{
+	apisixv1.SchemeHTTP:  80,
+	apisixv1.SchemeHTTPS: 443,
+	apisixv1.SchemeGRPC:  80,
+	apisixv1.SchemeGRPCS: 443,
+}
+
+// SchemeToPort scheme converts to the default port
+// ref https://github.com/apache/apisix/blob/master/apisix/upstream.lua#L167-L172
```
Comment (tao12345666333) [inline pkg/providers/utils/scheme.go:32]:
```suggestion
// ref https://github.com/apache/apisix/blob/c5fc10d9355a0c177a7532f01c77745ff0639a7f/apisix/upstream.lua#L167-L172
```
please use permalink

File: test/e2e/suite-features/external-sd.go [hunk 3]
```diff
@@ -23,10 +23,10 @@ import (
 	"os"
 	"time"
 
+	"github.com/apache/apisix-ingress-controller/pkg/config"
 	"github.com/onsi/ginkgo/v2"
 	"github.com/stretchr/testify/assert"
 
-	"github.com/apache/apisix-ingress-controller/pkg/config"
```
Comment (tao12345666333) [inline test/e2e/suite-features/external-sd.go:29]: this one don't need change.

File: test/e2e/suite-chore/admin_api.go [hunk 4]
```diff
@@ -18,10 +18,10 @@ package chore
 import (
 	"context"
 
+	v1 "github.com/apache/apisix-ingress-controller/pkg/types/apisix/v1"
 	"github.com/onsi/ginkgo/v2"
 	"github.com/stretchr/testify/assert"
 
-	v1 "github.com/apache/apisix-ingress-controller/pkg/types/apisix/v1"
```
Comment (tao12345666333) [inline test/e2e/suite-chore/admin_api.go:24]: ditto

File: pkg/providers/apisix/translation/apisix_upstream.go [hunk 5]
```diff
@@ -57,35 +55,24 @@ func (t *translator) translateService(namespace, svcName, subset, svcResolveGran
...
 		if node.Type == v2.ExternalTypeDomain {
-			arr := strings.Split(node.Name, ":")
 
 			weight := translation.DefaultWeight
 			if node.Weight != nil {
 				weight = *node.Weight
 			}
 			n := apisixv1.UpstreamNode{
-				Host:   arr[0],
+				Host:   node.Name,
```
Comment (tao12345666333) [inline pkg/providers/apisix/translation/apisix_upstream.go:68]: this means user only can use `domain`, right?

I think we should add some logic to make sure it is correct and also add test case to cover it.
For example, if the user fills in https://github.com , how should we handle it?

**Review by tao12345666333 — CHANGES_REQUESTED** [review]

File: pkg/providers/apisix/translation/apisix_upstream.go (hunk 5 above)
Comment (Gallardot) [reply pkg/providers/apisix/translation/apisix_upstream.go:68]: Yes, only domain can be used, or more precisely, the regular expression to match host.

```lua
local host_def_pat = "^\\*?[0-9a-zA-Z-._\\[\\]:]+$"
```

ref: https://github.com/apache/apisix/blob/c5fc10d9355a0c177a7532f01c77745ff0639a7f/apisix/schema_def.lua#L40

We can provide user-friendly error messages on ingress.

File: test/e2e/suite-features/external-sd.go (hunk 3 above)
Comment (Gallardot) [reply test/e2e/suite-features/external-sd.go:29]: Goland IDE formatted it automatically. 😂

File: pkg/providers/apisix/translation/apisix_upstream.go (hunk 5 above)
Comment (Gallardot) [reply pkg/providers/apisix/translation/apisix_upstream.go:76]: ref to apisix : https://github.com/apache/apisix/blob/c5fc10d9355a0c177a7532f01c77745ff0639a7f/apisix/schema_def.lua#L40
ref to k8s : https://github.com/kubernetes/kubernetes/blob/976a940f4a4e84fe814583848f97b9aafcdb083f/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L205

They define regex differently, but I think k8s's dns is more accurate.

Comment (tao12345666333) [reply pkg/providers/apisix/translation/apisix_upstream.go:76]: I think we can use k8s's rule

**Review by lingsamuel — APPROVED** [review]

**Review by tao12345666333 — APPROVED** [review]
LGTM

**Pull request merged**